### PR TITLE
feat(tappay-notify): endpoint for tappay backend notify

### LINKED ISSUE #769: [API] Webhook for Tappay
讓 Tappay 可以打我們的 endpoint

### PR DESCRIPTION
**Pull request by LittleWhiteYA**
Close #769 

## 這個 PR 是？

endpoint for tappay backend notify

開一個 endpoint，當 tappay 付費成功後，會打 request 到這個 endpoint 更新 PaymentRecord status

## 我應該從何看起？  

https://hackmd.io/khxI9FJCSW-1QG8mUo7fTA

https://docs.tappaysdk.com/tutorial/zh/back.html#backend-notify

## 若可以手動測試，要如何測試？ 

```
curl --location --request POST 'localhost:3000/tappay/notify' \
--header 'Content-Type: application/json' \
--data-raw '{
    "rec_trade_id": "xxx",
    "status": 0,
    "msg": "Success"
}'
```

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -73,6 +74,8 @@ app.use((req, res, next) => {
     next();
 });
 
+app.post("/tappay/notify", tappayNotifyRouter);
+
```
Comment (mark86092) [inline src/app.js:78]: I think we should add routing in 

https://github.com/goodjoblife/WorkTimeSurvey-backend/blob/0af1cf12c7f44a5ce2f98a435ddc7b668e955887/src/routes/index.js#L5-L14

File: src/libs/tappay.js [hunk 2]
```diff
@@ -7,6 +12,21 @@ const TAPPAY_API_MAPPING = {
 };
 const TAPPAY_API = TAPPAY_API_MAPPING[TAPPAY_ENV];
 
+const tappayNotifyRouter = wrap(async (req, res) => {
+    const { rec_trade_id, status, msg } = req.body;
```
Comment (mark86092) [inline src/libs/tappay.js:16]: Does tappay has any security check ?, or anyone can send fake post to our backend

File: src/libs/tappay.js [hunk 3]
```diff
@@ -7,6 +12,21 @@ const TAPPAY_API_MAPPING = {
...
+        await PaymentRecord.updateOne(
+            {
+                "tappaySnapshot.rec_trade_id": rec_trade_id,
+            },
+            { status: PAYMENT_RECORD_STATUS.Paid }
+        );
+    }
+
+    res.sendStatus(200);
+});
```
Comment (mark86092) [inline src/libs/tappay.js:28]: I would suggest put routing handler in `src/routes/`

File: src/libs/tappay.js (hunk 2 above)
Comment (barry800414) [reply src/libs/tappay.js:16]: I guess they have IP whitelist. But let me check with TapPay

File: src/routes/tappay/index.test.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    Paid,
+} = require("../../models/schemas/paymentRecordStatus");
+
+describe("/tappay/notify Test", () => {
+    const db = mongoose.connection;
+    const fake_rec_trade_id = "fake_rec_trade_id";
+    let payment_record_id = null;
+
+    before("Create test data", async () => {
+        const result = await db.collection("paymentRecords").insertOne({
```
Comment (mark86092) [inline src/routes/tappay/index.test.js:18]: use 
```js
const { PaymentRecord } = require("../../models");
await PaymentRecord.collection.insertOne({})
```

File: src/routes/tappay/index.test.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            })
+            .expect(200);
+        const paymentRecord = await db
+            .collection("paymentRecords")
+            .findOne({ _id: new ObjectId(payment_record_id) });
+        assert.propertyVal(paymentRecord, "status", Paid);
+    });
+
+    after(async () => {
+        await db.collection("paymentRecords").deleteMany({});
```
Comment (mark86092) [inline src/routes/tappay/index.test.js:59]: `PaymentRecord.collection`

File: src/libs/tappay.js (hunk 2 above)
Comment (barry800414) [reply src/libs/tappay.js]: I create an issue for addressing this issue https://github.com/goodjoblife/WorkTimeSurvey-backend/issues/810 
Let us discuss and do it after E2E connect with Tappay

File: src/routes/tappay/index.test.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                rec_trade_id: fake_rec_trade_id,
+                pay_info: {
+                    point: 0,
+                    masked_credit_card_number: "************2178",
+                    method: "CREDIT_CARD",
+                },
+            })
+            .expect(200);
+        const paymentRecord = await db
+            .collection("paymentRecords")
```
Comment (mark86092) [inline src/routes/tappay/index.test.js:54]: use `PaymentRecord`

**Review by mark86092 — APPROVED** [review]

**Pull request merged**